### DISCUSSION TITLE
[Native DLL] How to convert PascalABC.NET strings to UTF-8?

### DISCUSSION THREAD

**Discussion by ghost** [pascalabcnet/pascalabcnet#3010] [General]
The native library expects strings to be encoded in UTF-8. PascalABC.NET strings are native .NET strings which are encoded with UTF-16. My external function looks like this:

`function new_window(title: string): integer; external 'mydll.dll' name 'new_window';`

If I pass a PascalABC.NET string which contains Unicode characters to the function, what I see in the title of the window is not the Unicode characters but full of `?` characters.

**Comment by SunSerega** [comment]
Look at `System.Runtime.InteropServices.UnmanagedType.LPUTF8Str`

**Reply by ghost** [reply]
Please give me code. This is the first time I do pinvoke in PascalABC.NET. I don't know the syntax.

**Reply by SunSerega** [reply]
It's the exact same as in C#.
```
function new_window([MarshalAs(UnmanagedType.LPUTF8Str)] title: string): integer;
  external 'mydll.dll' name 'new_window';
```

**Reply by ghost** [reply]
Btw, I just read the document of pinvoke. Isn't CharSet Ansi (UTF-8) is the default? Why do it failed to work?

https://learn.microsoft.com/en-us/dotnet/standard/native-interop/charset

**Reply by SunSerega** [reply]
ANSI and UTF-8 are two different encodings. The first 128 code points always match, but in UTF-8 a single Unicode character can be represented by multiple bytes (multiple `char` items in a row), while in ANSI every character in 1-byte, and the top 128 code points are defined by system language - whatever you chose when first installing windows (and it doesn't change if you later change system language).

The link you posted says that UTF-8 is used by default on .Net Core. Meanwhile, PascalABC.Net currently plans to transition from .Net Framework to .Net Core, but there are still some things to figure out/reimplement to support .Net Core.

**Reply by SunSerega** [reply]
Anyway, relying on defaults in cases like these is generally not a very good idea. It's only valid while you are trying to make something simple, but the moment you want non-ASCII characters (more than the first 128 code points) - you better specify encoding explicitly.
